text/0715-construct-upgrade-support.md

```diff
@@ -0,0 +1,1274 @@
+is applied correctly to the migrated resources in the stack, especially for resources like DynamoDB tables.
+The process will block the migration if the `RETAIN` deletion policy is missing and ask users to deploy their
+V1 stack by setting `RETAIN` for deletion policy.
+
+We aim to improve this process by making this a streamlined process in the framework.
+During `cdk construct-upgrade` CLI command, if we detect that the retain policy is not enabled, we can
+do the following:
+
+Prompt the user with a message like:
+
+```sh
+Detected AWS::DynamoDB::Table with logical ID 'MyTable' does not have 'RETAIN' set as the deletion policy.
+CDK will deploy your stack with the deletion policy set to RETAIN on 'MyTable'.
+
+Would you like to proceed (y/n)?
+```
+
+If the user confirms by selecting "y", the migration will proceed to retrieve the deployed CloudFormation
+stack template, modify the `MyTable` resource in the template by setting `DeletionPolicy: Retain` and deploy
+the stack template. Once the deployment is successful, it will proceed with the usual flow to continue
+with validations as well as execution and deployment. This can be further automated with the `--require-approval`
+flag in the CLI command and if users choose `never` as the vaue, it will automatically set retain policy and
+deploy the change.
+
+#### Custom validations
+
+Custom validations ensure that the construct upgrade process adheres to best practices with custom
+configurations. For example, when migrating `Table` to `TableV2` for DynamoDB, the CDK team provides
+validations to ensure that custom resource deletions do not affect replica tables, that `TableName` and
+`SkipReplicaDeletion` properties are correctly set, and that `DeletionProtection` is enabled for safety.
+
+The `cdk diff` command will be extended to support a new `--custom-rule` option, which accepts a
+JavaScript file implementing a predefined `Rules` interface. These custom rule files allow both
+the CDK team and end-users to define and enforce additional rules during the diff process, enabling safer
+and more opinionated construct upgrade paths. When the `--custom-rule` option is specified, the CLI
+will `require()` the input file, instantiate the `Rules` object, and invoke the custom rules specified.
+An example is provided below for the custom rule set up and usage.
+
+```js
+export interface Rules {
+  /**
+   * The version of the Rules interface used. This will be used by
+   * the plug-in host to handle version changes.
+   */
+  version: '1';
+
+  /**
+   * When defined, this function is invoked right after the rules has been loaded,
+   * so that the plug-in is able to initialize itself and be picked by `cdk diff`
+   * command in the invocation.
+   */
+  init?: (host: IPluginHost) => void;
+}
+```
+
+An example of users' provided custom validation file is as follows:
+
+```ts
+import { IRules, Rules } from '@aws-cdk/cli-custom-rules';
+
+class DemoValidation {
```

Comment:
Shouldn't this implement some interface as well? how would the CLI know to call the `execute` method?